src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -4833,18 +4895,32 @@ std::optional<ProjectionCandidate> MergeTreeData::getQueryProcessingStageWithAgg
             }
         }
 
-        if (projection.type == ProjectionDescription::Type::Normal && (analysis_result.hasWhere() || analysis_result.hasPrewhere()))
+        if (projection.type == ProjectionDescription::Type::Normal)
         {
-            const auto & actions
-                = analysis_result.before_aggregation ? analysis_result.before_aggregation : analysis_result.before_order_by;
-            NameSet required_columns;
-            for (const auto & column : actions->getRequiredColumns())
-                required_columns.insert(column.name);
+            if (analysis_result.before_aggregation && analysis_result.optimize_aggregation_in_order)
+            {
+                for (const auto & key : keys)
+                {
+                    auto actions_dag = analysis_result.before_aggregation->clone();
+                    actions_dag->foldActionsByProjection({key}, sample_block_for_keys);
```

Comment:
May I ask to add a comment to `foldActionsByProjection`, `reorderAggregationKeysForProjection` and `addAggregatesViaProjection` methods of `ActionDAG`?

Reply:
Done